nncf/torch2/hook_executor/wrapper.py

```diff
@@ -0,0 +1,219 @@
+        if not isinstance(state, tuple):
+            raise TypeError("argument to __setstate__ must be a tuple")
+        if len(state) != 2:
+            raise TypeError(f"expected 2 items in state, got {len(state)}")
+        func, namespace = state
+        if not callable(func) or (namespace is not None and not isinstance(namespace, dict)):
+            raise TypeError("invalid partial state")
+
+        if namespace is None:
+            namespace = {}
+
+        self.__dict__ = namespace
+        self._func = func
+
+    @property
+    def func(self) -> MethodType:
+        return cast(MethodType, self._func)
+
+
+def wrap_model(model: nn.Module) -> nn.Module:
+    """
+    Wraps a nn.Module to inject custom behavior into the forward pass and replication process.
+
+    This function modifies the given model by:
+    1. Replacing the model's `forward` method with a wrapped version (`ForwardWithHooks`) that allows
+       additional hooks to be executed during the forward pass by using HookExecutorMode.
+    2. Wrapping the model's `_replicate_for_data_parallel` method with `ReplicateForDataParallel`,
+       which allows custom behavior when the model is replicated across multiple devices (e.g., for
+       data parallelism).
+    3. Adding a new module, `HookStorage`, to the model under the attribute `ATR_HOOK_STORAGE`.
+
+    :param model: The nn.Module to be wrapped.
+    :return: The modified model with the custom behavior injected.
+    """
+
+    if "forward" in model.__dict__:
+        raise nncf.InternalError("Wrapper does not supported models with overrided forward function")
+    model.forward = ForwardWithHooks(model.forward)
+    model._replicate_for_data_parallel = ReplicateForDataParallel(model._replicate_for_data_parallel)  # type: ignore
+    model.add_module(ATR_HOOK_STORAGE, HookStorage())
+    return model
+
+
+def is_wrapped(model: nn.Module) -> bool:
+    """
+    Checks if a given model has been wrapped by the `wrap_model` function.
+
+    :param model: The nn.Module to check.
+    :return: `True` if the model's `forward` method is an instance of `ForwardWithHooks`,
+        indicating that the model has been wrapped; `False` otherwise.
+    """
+    return isinstance(model.forward, ForwardWithHooks)
+
+
+def get_hook_storage(model: nn.Module) -> HookStorage:
+    """
+    Retrieves the `HookStorage` module from the given model.
+
+    This function accesses the model's attribute defined by `ATR_HOOK_STORAGE`
+    and returns the `HookStorage` module associated with it.
+
+
+    :param model: The PyTorch model from which to retrieve the `HookStorage`.
+    :return: The `HookStorage` module associated with the model.
+    """
+    storage = getattr(model, ATR_HOOK_STORAGE)
+    if storage is None:
+        raise nncf.InstallationError("Hook storage is not exist in the model")
+    return cast(HookStorage, getattr(model, ATR_HOOK_STORAGE))
+
+
+def insert_hook(
```

Comment:
PyTorch has the hook API https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#register_module_forward_hook. What do you think about reusing ideas from PyTorch to be friendly to PyTorch users? 

I mean the following ideas:
- `insert_hook` -> `register_pre_function_hook` and `register_post_function_hook` or you can use `insert` instead of `register`
- return `RemovableHandle`
- It looks like `group_name` is a hook name. I would suggest to think about removing `group_name` and letting users group hooks on their side.

Reply:
RemovableHandle